File: clients/client-dart/test/decoder/json_decoder_test.dart
import 'package:channel_sender_client/src/decoder/json_decoder.dart';
import 'package:test/test.dart';

void main() {
  group('Json decoding', () {
    test('should parse authOk event', () {
      const trama = '["", "", "AuthOk", ""]';

      final msg = JsonDecoder().decode(trama);
      expect(msg, isNotNull);
      expect(msg.messageId, equals(''));
      expect(msg.event, equals('AuthOk'));
      expect(msg.correlationId, equals(''));
      expect(msg.payload, equals(''));
    });

    test('should parse heartbeat event', () {
      const trama = '["", "1", ":hb", ""]';

      final msg = JsonDecoder().decode(trama);
      expect(msg, isNotNull);
      expect(msg.messageId, equals(''));
      expect(msg.event, equals(':hb'));
      expect(msg.correlationId, equals('1'));
      expect(msg.payload, equals(''));
    });

    test('should parse user event with json content', () {
      const trama =
          '["msg-id-0001","correlation-id-002","event.productCreated","{\\"code\\":\\"100\\", \\"title\\":\\"process after 5000ms\\", \\"detail\\":\\"some detail 89bd02d1da483efaa5389cbd4ca65bbd\\", \\"level\\":\\"info\\"}"]';

      final msg = JsonDecoder().decode(trama);

      expect(msg, isNotNull);
      expect(msg.messageId, equals('msg-id-0001'));
      expect(msg.correlationId, equals('correlation-id-002'));
      expect(msg.event, equals('event.productCreated'));
      expect(
        msg.payload,
        equals(
            '{"code":"100", "title":"process after 5000ms", "detail":"some detail 89bd02d1da483efaa5389cbd4ca65bbd", "level":"info"}'),
      );
    });

    test('should parse user event with no json content', () {
      const trama =
          '["msg-id-0003","correlation-id-004","event.productCreated","Hello World"]';

      final msg = JsonDecoder().decode(trama);

      expect(msg, isNotNull);
      expect(msg.messageId, equals('msg-id-0003'));
      expect(msg.correlationId, equals('correlation-id-004'));
      expect(msg.event, equals('event.productCreated'));
      expect(msg.payload, equals('Hello World'));
    });

    test('should parse user event with no payload', () {
      const trama =
          '["msg-id-0003","correlation-id-004","event.productCreated",""]';

      final msg = JsonDecoder().decode(trama);

      expect(msg, isNotNull);
      expect(msg.messageId, equals('msg-id-0003'));
      expect(msg.correlationId, equals('correlation-id-004'));
      expect(msg.event, equals('event.productCreated'));
      expect(msg.payload, equals(''));
    });

    test('should parse user event with weird payload', () {
      const trama =
          '["msg-id-0003","correlation-id-004","event.productCreated",".,.,.,%%#@"]';

      final msg = JsonDecoder().decode(trama);

      expect(msg, isNotNull);
      expect(msg.messageId, equals('msg-id-0003'));
      expect(msg.correlationId, equals('correlation-id-004'));
      expect(msg.event, equals('event.productCreated'));
      expect(msg.payload, equals('.,.,.,%%#@'));
    });
  });
}

File: clients/client-js/test/decoder/json-decoder.test.ts
import * as chai from 'chai';

import { MessageDecoder, JsonDecoder } from "../../src/decoder/index.js";
import { ChannelMessage } from "../../src/channel-message.js";

const assert = chai.assert;
describe('Json Serializer Tests', function () {

    const serializer: MessageDecoder = new JsonDecoder()

    it('Should decode basic string payload', () => {
        const payload = "[\"ids332msg1\", \"\", \"person.registered\", \"someData\"]";
        const event = new MessageEvent('test', {
            data: payload
        });
        const message = serializer.decode(event);
        assert.deepEqual(message, new ChannelMessage("ids332msg1", "person.registered", "", "someData"));
    });

    it('Should decode basic sse string payload', () => {
        const sse_event = {
            id: undefined,
            data: '["aa6db9ef-bbad-41dc-9099-4e28bbb9c37b","7595b125-e503-4a0b-a4d5-c92edd5591c2","businessEvent",{"code":"100","title":"process after 500","severity":"INFO","detail":"some detail 014438cd-ed20-488a-8237-c8cb9e4d18c5"}]',
            event: 'message',
            retry: undefined
        }
        const message = serializer.decode_sse(sse_event.data);
        assert.deepEqual(message, new ChannelMessage("aa6db9ef-bbad-41dc-9099-4e28bbb9c37b", "businessEvent", "7595b125-e503-4a0b-a4d5-c92edd5591c2", { "code": "100", "title": "process after 500", "severity": "INFO", "detail": "some detail 014438cd-ed20-488a-8237-c8cb9e4d18c5" }));
    });
});
